File: subject_teacher/drive/schemas.py
"""Pydantic schemas for files stored in Google Drive appDataFolder."""
from __future__ import annotations

import re
from enum import Enum
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator
from pydantic.alias_generators import to_camel

import regions

SCHEMA_VERSION = 1

_CLASS_KEY_RE = re.compile(r"^\d+-[^\t\r\n]+$")
_MONTH_RE = re.compile(r"^\d{4}-(0[1-9]|1[0-2])$")


def _normalize_class_no(value: object) -> str:
    if value is None:
        raise ValueError("classNo must not be blank")
    text = str(value).strip()
    if not text:
        raise ValueError("classNo must not be blank")
    if any(char in text for char in "\t\r\n"):
        raise ValueError("classNo must not contain tab or newline characters")
    return text


def _normalize_grade(value: object) -> int:
    if isinstance(value, str):
        match = re.search(r"\d+", value)
        if match:
            return int(match.group(0))
    return value


class _Base(BaseModel):
    """Base model with camelCase aliases and forgiving extra fields."""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="ignore",
    )


class Semester(_Base):
    year: int = Field(ge=2024, le=2100)
    term: Literal[1, 2]


class AssignedLesson(_Base):
    grade: Annotated[int, Field(ge=1, le=3)]
    class_no: Annotated[str, Field(min_length=1, max_length=40)]
    subject_name: str
    neis_subject_label: str = ""
    subject_aliases: list[str] = Field(default_factory=list)

    @field_validator("grade", mode="before")
    @classmethod
    def validate_grade(cls, value: object) -> int:
        return _normalize_grade(value)

    @field_validator("class_no", mode="before")
    @classmethod
    def validate_class_no(cls, value: object) -> str:
        return _normalize_class_no(value)


class Settings(_Base):
    schema_version: int = Field(default=SCHEMA_VERSION)
    teacher_name: str
    school_name: str
    region: str
    semester: Semester
    close_by_default: bool = False
    timetable_mode: Literal["manual", "neis"] = "neis"
    assigned_lessons: list[AssignedLesson] = Field(default_factory=list)
    updated_at: str

    @field_validator("region")
    @classmethod
    def validate_region(cls, value: str) -> str:
        if value not in regions.REGIONS:
            raise ValueError(f"unknown region: {value}")
        return value


class TimetableSlot(_Base):
    id: str
    day_of_week: Annotated[int, Field(ge=1, le=5)]
    period: Annotated[int, Field(ge=1, le=7)]
    grade: Annotated[int, Field(ge=1, le=3)]
    class_no: Annotated[str, Field(min_length=1, max_length=40)]
    subject_name: str
    neis_subject_label: str

    @field_validator("class_no", mode="before")
    @classmethod
    def validate_class_no(cls, value: object) -> str:
        return _normalize_class_no(value)


class Timetable(_Base):
    schema_version: int = Field(default=SCHEMA_VERSION)
    effective_from: str
    slots: list[TimetableSlot]


class StudentEntry(_Base):
    number: Annotated[int, Field(ge=1, le=99)]
    name: str = ""


class Students(_Base):
    schema_version: int = Field(default=SCHEMA_VERSION)
    classes: dict[str, list[StudentEntry]]

    @field_validator("classes")
    @classmethod
    def validate_classes(cls, value: dict[str, list[StudentEntry]]) -> dict[str, list[StudentEntry]]:
        for key in value:
            if not _CLASS_KEY_RE.match(key) or not key.split("-", 1)[1].strip():
                raise ValueError(f"class key must match 'grade-classNo', got {key!r}")
        return value


class MarkType(str, Enum):
    ABSENT = "absent"
    EXCUSED = "excused"


class Absence(_Base):
    student_number: Annotated[int, Field(ge=1, le=99)]
    mark_type: MarkType
    note: str = ""


class SlotAttendance(_Base):
    absences: list[Absence]
    checked_at: str
    source: Literal["mobile", "pc"] = "mobile"
    synced_to_neis: bool = False
    closed_on_neis: bool = False


class DayAttendance(_Base):
    """Placeholder model kept for import compatibility in early plan steps."""


class MonthlyAttendance(_Base):
    schema_version: int = Field(default=SCHEMA_VERSION)
    month: str
    records: dict[str, dict[str, SlotAttendance]]

    @field_validator("month")
    @classmethod
    def validate_month(cls, value: str) -> str:
        if not _MONTH_RE.match(value):
            raise ValueError(f"month must be 'YYYY-MM', got {value!r}")
        return value


def numbers_only(students: "Students") -> "Students":
    """Return a copy of the roster with all names stripped (numbers only)."""
    return Students(
        schemaVersion=students.schema_version,
        classes={
            key: [StudentEntry(number=entry.number) for entry in entries]
            for key, entries in students.classes.items()
        },
    )

File: subject_teacher_pwa/src/lib/schemas.ts
import { z } from "zod";

export const SCHEMA_VERSION = 1;

const schemaVersion = z.literal(SCHEMA_VERSION);
const isoMonth = z.string().regex(/^\d{4}-(0[1-9]|1[0-2])$/);
const isoDate = z.string().regex(/^\d{4}-\d{2}-\d{2}$/);
const classKey = z.string().regex(/^\d+-[^\t\r\n]+$/);

const classNo = z.preprocess((value) => {
  if (value === null || value === undefined) return value;
  return String(value).trim();
}, z.string().min(1).max(40).refine((value) => !/[\t\r\n]/.test(value), {
  message: "classNo must not contain tabs or line breaks",
}));

export const SemesterSchema = z.object({
  year: z.number().int().min(2024).max(2100),
  term: z.union([z.literal(1), z.literal(2)]),
});

export const SettingsSchema = z.object({
  schemaVersion: schemaVersion.default(SCHEMA_VERSION),
  teacherName: z.string(),
  schoolName: z.string(),
  region: z.string().min(1),
  semester: SemesterSchema,
  closeByDefault: z.boolean().default(false),
  updatedAt: z.string().min(1),
});

export const TimetableSlotSchema = z.object({
  id: z.string().min(1),
  dayOfWeek: z.number().int().min(1).max(5),
  period: z.number().int().min(1).max(7),
  grade: z.number().int().min(1).max(3),
  classNo,
  subjectName: z.string().min(1),
  neisSubjectLabel: z.string().min(1),
});

export const TimetableSchema = z.object({
  schemaVersion: schemaVersion.default(SCHEMA_VERSION),
  effectiveFrom: isoDate,
  slots: z.array(TimetableSlotSchema),
});

export const StudentEntrySchema = z.object({
  number: z.number().int().min(1).max(99),
  name: z.string().default(""),
});

export const StudentsSchema = z.object({
  schemaVersion: schemaVersion.default(SCHEMA_VERSION),
  classes: z.record(classKey, z.array(StudentEntrySchema)),
});

export const MarkTypeSchema = z.union([z.literal("absent"), z.literal("excused")]);

export const AbsenceSchema = z.object({
  studentNumber: z.number().int().min(1).max(99),
  markType: MarkTypeSchema,
  note: z.string().default(""),
});

export const SlotAttendanceSchema = z.object({
  absences: z.array(AbsenceSchema),
  checkedAt: z.string().min(1),
  source: z.union([z.literal("mobile"), z.literal("pc")]).default("mobile"),
  syncedToNeis: z.boolean().default(false),
  closedOnNeis: z.boolean().default(false),
});

export const MonthlyAttendanceSchema = z.object({
  schemaVersion: schemaVersion.default(SCHEMA_VERSION),
  month: isoMonth,
  records: z.record(isoDate, z.record(z.string().min(1), SlotAttendanceSchema)),
});

export type Settings = z.infer<typeof SettingsSchema>;
export type TimetableSlot = z.infer<typeof TimetableSlotSchema>;
export type Timetable = z.infer<typeof TimetableSchema>;
export type StudentEntry = z.infer<typeof StudentEntrySchema>;
export type Students = z.infer<typeof StudentsSchema>;
export type MarkType = z.infer<typeof MarkTypeSchema>;
export type Absence = z.infer<typeof AbsenceSchema>;
export type SlotAttendance = z.infer<typeof SlotAttendanceSchema>;
export type MonthlyAttendance = z.infer<typeof MonthlyAttendanceSchema>;
